Add arithmetic circuit evaluation op

Al-Kindi-0 opened the pull request:
This PR introduces the ACE chiplet into the VM and it has two components:

1. The main logic for handling arithmetic circuits targeted for ACE as per its specs and is contained `/miden-vm/processor/src/chiplets/ace/mod.rs`. This includes the core of trace generation associated to the chiplet.
2. The logic for integrating the ACE chiplet into the VM. This includes adding an instruction and integrating the chiplet into the chiplets trace.

adr1anh commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,286 @@
+    fn num_fractions(&self) -> usize {
+        self.sections
+            .iter()
+            .fold(0, |acc, term| acc + 2 * term.num_vars + 3 * term.num_evals) as usize
```
Nit: `.iter().map(...).sum()`

Al-Kindi-0 replied:
Thanks, changed now

adr1anh reviewed (commented):
Not a full review, but amazing work! The only thing missing would be to move the circuit construction logic to `test`

Al-Kindi-0 commented on air/src/trace/main_trace.rs:
```diff
@@ -411,41 +416,149 @@ impl MainTrace {
+    }
+
+    pub fn chiplet_ace_start_selector(&self, i: RowIndex) -> Felt {
+        self.columns.get_column(CHIPLETS_OFFSET + 4)[i]
```
Makes sense
Done!

Al-Kindi-0 commented on air/src/trace/chiplets/ace.rs:
```diff
@@ -0,0 +1 @@
+
```
That was the plan and the reason for the module creation but for some reason forgot to do it.
Done now

Al-Kindi-0 commented on processor/src/chiplets/ace/encoded_circuit.rs:
```diff
@@ -0,0 +1,84 @@
+
+    /// Returns the digest of the circuit including a header
+    pub fn circuit_hash<H: ElementHasher<BaseField = Felt>>(&self) -> H::Digest {
+        todo!()
```
Removed for now, we can add back if/when needed

Al-Kindi-0 commented:
> "... The only thing missing would be to move the circuit construction logic to `test`"

You mean `encoded_circuit.rs`?

Al-Kindi-0 commented on processor/src/chiplets/ace/tests/encoder.rs:
```diff
@@ -0,0 +1,146 @@
+    fn padded(&self) -> Self {
+        // Inputs are padded to next multiple of 2 so they can be word-aligned, since each word
+        // contains two inputs.
+        // TODO(@adr1anh): does it makes sense to double-word align?
```
Maybe not, but we can keep it for now and fix it if needed when we can run the end-to-end test

Al-Kindi-0 commented on processor/src/chiplets/aux_trace/mod.rs:
```diff
@@ -2,47 +2,62 @@ use alloc::vec::Vec;
+    ///
+    /// 1. a bus column `b_chip` describing requests made by the stack and decoder and responses
+    ///    received from the chiplets in the Chiplets module,
+    /// 2. a column acting as both virtual tables, one for the sibling table used by the hasher
+    ///    chiplet and the other for the kernel procedure table, and as a bus between the memory
+    ///    chiplet and the ACE chiplet.
```
note that the virtual table (bus) column is now expanded to handle the messages between ACE and memory. We should probably include this in the updated docs.

Al-Kindi-0 reviewed (commented):
Awesome work @adr1anh !
I have made some minor updates related to the formatting of the docs, but other than that all looks good to me.

adr1anh commented on air/src/trace/main_trace.rs:
```diff
@@ -12,6 +12,11 @@ use super::{
+            CLK_IDX, CTX_IDX, EVAL_OP_IDX, ID_0_IDX, ID_1_IDX, ID_2_IDX, M_0_IDX, M_1_IDX,
+            NUM_ACE_EXTERNAL_SELECTORS, PTR_IDX, SELECTOR_BLOCK_IDX, SELECTOR_START_IDX, V_0_0_IDX,
+            V_0_1_IDX, V_1_0_IDX, V_1_1_IDX, V_2_0_IDX, V_2_1_IDX,
+        },
```
Goes along with change below
```suggestion
        ace::{
            CLK_IDX, CTX_IDX, EVAL_OP_IDX, ID_0_IDX, ID_1_IDX, ID_2_IDX, M_0_IDX, M_1_IDX, PTR_IDX,
            READ_NUM_EVAL_IDX, SELECTOR_BLOCK_IDX, SELECTOR_START_IDX, V_0_0_IDX, V_0_1_IDX,
            V_1_0_IDX, V_1_1_IDX, V_2_0_IDX, V_2_1_IDX,
        },
```

adr1anh commented on air/src/trace/main_trace.rs:
```diff
@@ -411,41 +421,158 @@ impl MainTrace {
+    pub fn chiplet_ace_v_0_1(&self, i: RowIndex) -> Felt {
+        self.columns
+            .get_column(CHIPLETS_OFFSET + NUM_ACE_EXTERNAL_SELECTORS + V_0_1_IDX)[i]
+    }
+    pub fn chiplet_ace_wire_0(&self, i: RowIndex) -> [Felt; 3] {
```
Adds a new line

adr1anh commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,292 @@
+#![allow(dead_code)]
+
+use alloc::{collections::BTreeMap, vec::Vec};
+
+use miden_air::{
+    RowIndex,
+    trace::{chiplets::ace::ACE_CHIPLET_NUM_COLS, main_trace::MainTrace},
+};
+use vm_core::{FieldElement, ZERO, mast::MastNodeExt};
+
+use crate::{
+    ContextId, ExecutionError, Felt, QuadFelt,
+    chiplets::{ace::trace::EvaluationContext, memory::Memory},
+    errors::{AceError, ErrorContext},
+    trace::TraceFragment,
+};
+
+mod trace;
+
+mod encoded_circuit;
+#[cfg(test)]
+mod tests;
```
Not entirely sure what the conventions are around imports, is it just what `rustfmt` does? I can update also to put them all in one block.

Al-Kindi-0 replied:
I think mostly following `rustfmt`

adr1anh commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,292 @@
+#![allow(dead_code)]
+
+use alloc::{collections::BTreeMap, vec::Vec};
+
+use miden_air::{
+    RowIndex,
+    trace::{chiplets::ace::ACE_CHIPLET_NUM_COLS, main_trace::MainTrace},
+};
+use vm_core::{FieldElement, ZERO, mast::MastNodeExt};
+
+use crate::{
+    ContextId, ExecutionError, Felt, QuadFelt,
+    chiplets::{ace::trace::EvaluationContext, memory::Memory},
+    errors::{AceError, ErrorContext},
+    trace::TraceFragment,
+};
+
+mod trace;
+
+mod encoded_circuit;
+#[cfg(test)]
+mod tests;
```
This is what it gives if I format it in a single block 
```suggestion
#![allow(dead_code)]

use crate::{
    ContextId, ExecutionError, Felt, QuadFelt,
    chiplets::{ace::trace::EvaluationContext, memory::Memory},
    errors::{AceError, ErrorContext},
    trace::TraceFragment,
};
use alloc::{collections::BTreeMap, vec::Vec};
use miden_air::{
    RowIndex,
    trace::{chiplets::ace::ACE_CHIPLET_NUM_COLS, main_trace::MainTrace},
};
use vm_core::{FieldElement, ZERO, mast::MastNodeExt};

mod encoded_circuit;
mod trace;

#[cfg(test)]
mod tests;
```

Al-Kindi-0 replied:
I have tried to tidy things a bit, let me know if there is anything we can do better

adr1anh commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,292 @@
+
+impl Ace {
+    pub(crate) fn trace_len(&self) -> usize {
+        self.circuit_evaluations.iter().fold(0, |acc, term| acc + term.1.num_rows())
```
Nit:
```suggestion
        self.circuit_evaluations.values().map(|eval_ctx| eval_ctx.num_rows()).sum()
```

Al-Kindi-0 replied:
Added, thank you!

adr1anh commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,292 @@
+            num_evals: eval_ctx.num_eval_rows(),
+        }
+    }
+}
+#[derive(Debug, Default)]
+pub struct AceHints {
```
Nit: new line 
```suggestion
}

#[derive(Debug, Default)]
pub struct AceHints {
```

adr1anh commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,292 @@
+            let clk = section.clk();
+            let ctx = section.ctx();
+
+            let values = &mut total_values[values_offset
+                ..values_offset + NUM_ACE_LOGUP_FRACTIONS_READ * section.num_vars() as usize];
+            let inv_values = &mut total_inv_values[values_offset
+                ..values_offset + NUM_ACE_LOGUP_FRACTIONS_READ * section.num_vars() as usize];
```
Small nit, but maybe not much clearer (up to you)
```suggestion
            let values_range =
                range(values_offset, NUM_ACE_LOGUP_FRACTIONS_READ * section.num_vars() as usize);
            let values = &mut total_values[values_range];
            let inv_values_range =
                range(values_offset, NUM_ACE_LOGUP_FRACTIONS_READ * section.num_vars() as usize);
            let inv_values = &mut total_inv_values[inv_values_range];
```

adr1anh reviewed (commented):
More thorough review, mostly small nits so feel free to ignore the suggestions

Al-Kindi-0 commented:
> More thorough review, mostly small nits so feel free to ignore the suggestions

Addressed and implemented most comments.

As for the imports, I think the convention is to keep the `alloc::*` the rest is to use whatever `rustfmt` suggests, but others can correct me.

Al-Kindi-0 commented on processor/src/operations/circuit_eval.rs:
```diff
@@ -0,0 +1,46 @@
+    /// 3. the number of rows in the `EVAL` section,
+    ///
+    /// Stack transition:
+    /// [ptr, num_read_rows, num_eval_rows, ...] -> [...]
```
Added, thanks!

Al-Kindi-0 commented on processor/src/chiplets/mod.rs:
```diff
@@ -33,91 +37,112 @@ pub(crate) use aux_trace::AuxTraceBuilder;
 ///   first rows of the trace up to the length of the hasher `trace_len`.
 ///   - column 0: selector column with values set to ZERO
 ///   - columns 1-16: execution trace of hash chiplet
 ///   - column 17: unused column padded with ZERO
```
Thanks for catching it

Al-Kindi-0 commented on air/src/trace/main_trace.rs:
```diff
@@ -411,41 +421,158 @@ impl MainTrace {
+    pub fn chiplet_ace_eval_op(&self, i: RowIndex) -> Felt {
+        self.columns
+            .get_column(CHIPLETS_OFFSET + NUM_ACE_EXTERNAL_SELECTORS + EVAL_OP_IDX)[i]
+    }
+
+    pub fn chiplet_ace_id_0(&self, i: RowIndex) -> Felt {
```
Makes sense

Al-Kindi-0 commented on processor/src/chiplets/aux_trace/bus/memory.rs:
```diff
@@ -24,6 +24,74 @@ const FOUR: Felt = Felt::new(4);
+
+    let id_0 = main_trace.chiplet_ace_id_1(row);
+    let id_1 = main_trace.chiplet_ace_id_2(row);
+    let element = id_0 + id_1 * Felt::new(1 << 30) + (element + ONE) * Felt::new(1 << 60);
```
Done!

Al-Kindi-0 commented on processor/src/chiplets/aux_trace/mod.rs:
```diff
@@ -2,47 +2,62 @@ use alloc::vec::Vec;
         &self,
         main_trace: &MainTrace,
         rand_elements: &[E],
     ) -> Vec<Vec<E>> {
```
Good idea, done!

Al-Kindi-0 commented on processor/src/chiplets/aux_trace/mod.rs:
```diff
@@ -2,47 +2,62 @@ use alloc::vec::Vec;
+    /// 2. a column acting as both virtual tables, one for the sibling table used by the hasher
+    ///    chiplet and the other for the kernel procedure table, and as a bus between the memory
+    ///    chiplet and the ACE chiplet.
+    /// 3. a column used as bus to wire the gates of the ACE chiplet.
```
Edited a bit and add it

paracetamolo commented on stdlib/tests/crypto/arithmetic_circuit_eval.rs:
```diff
@@ -0,0 +1,152 @@
+}
+
+#[test]
+fn arithmetic_circuit_eval() {
```
These two functions are almost identical, they can be refactored to share most code.

Al-Kindi-0 replied:
Indeed, collapsed both into one

paracetamolo reviewed (commented):
Just a code style review. For the next step it would be nice to make sure that all tests in the AirScript ACE backend can be run with the VM.

Al-Kindi-0 commented on air/src/trace/chiplets/ace.rs:
```diff
@@ -0,0 +1,75 @@
+/// Offset of the `ID1` wire used when encoding an ACE instruction.
+pub const ACE_INSTRUCTION_ID1_OFFSET: Felt = Felt::new(1 << 30);
+
+/// Offset of the `ID1` wire used when encoding an ACE instruction.
```
Good catch, fixed

Al-Kindi-0 commented on processor/src/chiplets/ace/encoded_circuit.rs:
```diff
@@ -0,0 +1,77 @@
+/// constants and instructions.
+#[derive(Clone, Debug)]
+pub struct EncodedCircuit {
+    pub num_vars: usize,
```
added

Al-Kindi-0 commented on processor/src/chiplets/ace/encoded_circuit.rs:
```diff
@@ -0,0 +1,77 @@
+
+    /// Computes the hash of all circuit constants and instructions.
+    fn raw_circuit_hash<H: ElementHasher<BaseField = Felt>>(&self) -> H::Digest {
+        debug_assert_eq!(self.encoded_circuit.len() % 8, 0);
```
moved to the constructor

Al-Kindi-0 commented on processor/src/chiplets/ace/encoded_circuit.rs:
```diff
@@ -0,0 +1,77 @@
+
+    /// Returns the number of constants in the circuit.
+    pub fn num_constants(&self) -> usize {
+        (self.encoded_circuit.len() - self.num_eval) / 2
```
Yes, the encoded circuit is a `Vec<Felt>` and constants are extension field elements and hence the division

Al-Kindi-0 commented on processor/src/chiplets/ace/encoded_circuit.rs:
```diff
@@ -0,0 +1,77 @@
+
+    /// Returns the number of inputs in the circuit.
+    pub fn num_inputs(&self) -> usize {
+        self.num_vars - self.num_constants()
```
Inputs are extension field elements but since we normalized by 2 above, the subtraction is a subtraction of numbers of extension field elements.

Al-Kindi-0 commented on processor/src/operations/horner_ops.rs:
```diff
@@ -453,6 +454,7 @@ mod tests {
     }
 
     #[test]
+    #[ignore]
```
Good point, removed the ignore

plafer commented on processor/src/chiplets/ace/instruction.rs:
```diff
@@ -0,0 +1,48 @@
+}
+
+/// Given a `Felt`, tries to recover the components `id_l, id_r, op`.
+pub fn decode_instruction(instruction: Felt) -> Option<(u32, u32, Op)> {
```
What's the difference between a node, an instruction and an operation? i.e. is the "instruction" the best name here, or should it be node?

Al-Kindi-0 replied:
An instruction encodes the identifiers of the incoming wires and the operation, which can be an add, sub, or a mul

plafer commented on processor/src/chiplets/ace/instruction.rs:
```diff
@@ -0,0 +1,48 @@
+/// `uint`.
+pub const MAX_ID: u32 = (1 << ID_BITS) - 1;
+
+/// Arithmetic operation applied to two nodes in the evaluation graph.
```
nit: Technically, shouldn't this say "applied to two wires" instead of "nodes"? Since the result of an operation is encoded in a wire.

Not so much trying to be picky than making sure I understand correctly how everything fits together (and making sure comments don't cause confusion for future readers).

Al-Kindi-0 replied:
Yeah, that's a bit confusing, changed it to `wire`

plafer reviewed (commented):
Left a bunch of nits/clarifications/docstring requests. But overall looks great!

Al-Kindi-0 commented on processor/src/chiplets/aux_trace/bus/memory.rs:
```diff
@@ -24,6 +27,75 @@ const FOUR: Felt = Felt::new(4);
+    let value = message.value(alphas);
+
+    #[cfg(any(test, feature = "bus-debugger"))]
+    _debugger.add_request(Box::new(message), alphas);
```
#1780

Al-Kindi-0 commented on processor/src/chiplets/aux_trace/mod.rs:
```diff
@@ -2,47 +2,63 @@ use alloc::vec::Vec;
+        let wiring_bus_builder = WiringBusBuilder::new(&self.ace_hints);
         let t_chip = v_table_col_builder.build_aux_column(main_trace, rand_elements);
         let b_chip = bus_col_builder.build_aux_column(main_trace, rand_elements);
+        let wiring_bus = wiring_bus_builder.build_aux_column(main_trace, rand_elements);
```
The is already done in the `build_aux_column()` method in  a stronger manner

Al-Kindi-0 commented on stdlib/tests/crypto/stark/mod.rs:
```diff
@@ -12,8 +12,11 @@ use verifier_recursive::{VerifierData, generate_advice_inputs};
 // Note: Changes to Miden VM may cause this test to fail when some of the assumptions documented
 // in `stdlib/asm/crypto/stark/verifier.masm` are violated.
 #[rstest]
+#[ignore = "fixed-in-next-pr"]
 #[case(None)]
+#[ignore = "fixed-in-next-pr"]
 #[case(Some(KERNEL_ODD_NUM_PROC))]
+#[ignore = "fixed-in-next-pr"]
```
done

Al-Kindi-0 commented on processor/src/errors.rs:
```diff
@@ -137,6 +137,8 @@ pub enum ExecutionError {
       hex = to_hex(.0.as_bytes())
     )]
     SyscallTargetNotInKernel(Digest),
+    #[error("failed to execute arithmetic circuit evaluation operation: {0}")]
+    AceError(AceError),
```
Added, let me know if it is uniform with the rest

Al-Kindi-0 commented on processor/src/operations/mod.rs:
```diff
@@ -170,9 +171,12 @@ impl Process {
             Operation::MpVerify(err_code) => self.op_mpverify(err_code, host)?,
             Operation::MrUpdate => self.op_mrupdate(host)?,
+
+            // ----- STARK proof verification -----------------------------------------------------
```
For the time being that's its only use and I don't see any other in the short to medium term

Al-Kindi-0 commented on processor/src/chiplets/ace/trace.rs:
```diff
@@ -0,0 +1,317 @@
+        self.col_w1.push(id_1, v_1);
+
+        // Add pointer to trace, and return the next one
+        const PTR_OFFSET: Felt = Felt::new(4);
+        self.col_ptr.push(ptr);
+        Ok(ptr + PTR_OFFSET)
```
Agreed, moved the logic outside

Al-Kindi-0 commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,293 @@
+    let num_eval = num_eval.as_int();
+
+    // Ensure vars and instructions are word-aligned and non-empty
+    if num_vars % 2 != 0 || num_vars == 0 {
+        return Err(ExecutionError::AceError(AceError::NumVarIsNotWordAlignedOrIsEmpty(num_vars)));
+    }
+    if num_eval % 4 != 0 || num_eval == 0 {
+        return Err(ExecutionError::AceError(AceError::NumEvalIsNotWordAlignedOrIsEmpty(num_eval)));
+    }
```
Added docs as you suggested

Al-Kindi-0 commented on processor/src/chiplets/ace/trace.rs:
```diff
@@ -0,0 +1,317 @@
+    col_ptr: Vec<Felt>,
+    col_op: Vec<Felt>,
+
+    col_w0: WireColumn,
+    col_w1: WireColumn,
+    col_w2: WireColumn,
```
Changed

Al-Kindi-0 commented on processor/src/chiplets/ace/trace.rs:
```diff
@@ -0,0 +1,317 @@
+///
+/// The set of nodes is used to fill the ACE chiplet trace.
+#[derive(Debug)]
+pub struct EvaluationContext {
```
I agree!
I have changed it to `CircuitEvaluation` as I feel it is slightly better than `EvaluatedCircuit` because it captures a process rather than a result, but I admit this is probably intangible

Al-Kindi-0 commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,293 @@
+}
+
+#[derive(Debug, Default)]
+pub struct AceHints {
```
Added a comment to explain that, with the hints, the logic for building the auxiliary trace becomes easier

Al-Kindi-0 commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,293 @@
+}
+
+#[derive(Debug, Default)]
+pub struct AceSection {
```
Changed

Al-Kindi-0 commented on processor/src/chiplets/ace/trace.rs:
```diff
@@ -0,0 +1,317 @@
+    }
+}
+
+/// Set of columns for a given wire containing `[id, v_0, v_1]`
```
Added a modified docstring to explain a bit more

Al-Kindi-0 commented on processor/src/chiplets/ace/trace.rs:
```diff
@@ -0,0 +1,317 @@
+        columns[CLK_IDX][offset..offset + num_rows].fill(clt_felt);
+
+        // Fill n_eval which is constant across the read block
+        let n_eval_felt = Felt::from(self.num_eval_rows - 1);
```
Indeed, the name is not congruent. I changed it to `eval_section_first_idx` to better capture that we are talking about indices which start at 0

Al-Kindi-0 commented on processor/src/chiplets/ace/trace.rs:
```diff
@@ -0,0 +1,317 @@
+}
+
+#[derive(Debug)]
+struct WireBus {
```
Added a docstring to explain further

Al-Kindi-0 commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,293 @@
+pub use trace::{NUM_ACE_LOGUP_FRACTIONS_EVAL, NUM_ACE_LOGUP_FRACTIONS_READ};
+
+#[derive(Debug, Default)]
+pub struct Ace {
```
Added

plafer commented on processor/src/chiplets/ace/mod.rs:
```diff
@@ -0,0 +1,345 @@
+/// to evaluating some multi-variate polynomial at a tuple representing the input.
+///
+/// During the course of the VM execution, we keep track of all calls to the ACE chiplet in an
+/// `EvaluationContext` per call. This is then used to generate the full trace of the ACE chiplet.
```
```suggestion
/// [`CircuitEvaluation`] per call. This is then used to generate the full trace of the ACE chiplet.
```

`EvaluationContext` no longer exists

plafer reviewed (approved):
LGTM! Although note that my review was somewhat superficial given that it was thoroughly reviewed by @adr1anh and @paracetamolo
